// @tanstack/react-start-plugin@1.121.16/dist/cjs/index.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const viteReact = require("@vitejs/plugin-react");
const startPluginCore = require("@tanstack/start-plugin-core");
const schema = require("./schema.cjs");
function TanStackStartVitePlugin(opts) {
  const options = schema.getTanStackStartOptions(opts);
  return [
    startPluginCore.TanStackStartVitePluginCore(
      {
        framework: "react",
        getVirtualServerRootHandler(ctx) {
          return `
import { toWebRequest, defineEventHandler } from '@tanstack/react-start/server';
import serverEntry from '${ctx.serverEntryFilepath}';

export default defineEventHandler(function(event) {
  const request = toWebRequest(event);
  return serverEntry({ request });
});`;
        },
        getVirtualClientEntry(ctx) {
          return `
import { StrictMode, startTransition } from 'react';
import { hydrateRoot } from 'react-dom/client';
import { StartClient } from '@tanstack/react-start';
import { createRouter } from '${ctx.routerFilepath}';

const router = createRouter();

startTransition(() => {
  hydrateRoot(
    document,
    <StrictMode>
      <StartClient router={router} />
    </StrictMode>
  );
});`;
        },
        getVirtualServerEntry(ctx) {
          return `
import { createStartHandler, defaultStreamHandler } from '@tanstack/react-start/server';
import { createRouter } from '${ctx.routerFilepath}';

export default createStartHandler({
  createRouter,
})(defaultStreamHandler);`;
        }
      },
      options
    ),
    viteReact(options.react)
  ];
}
exports.TanStackStartVitePlugin = TanStackStartVitePlugin;
exports.tanstackStart = TanStackStartVitePlugin;
