// @tanstack/start-server-core@1.121.16/dist/esm/router-manifest.js
import { rootRouteId, joinPaths } from "@tanstack/router-core";
import { VIRTUAL_MODULES } from "./virtual-modules.js";
import { loadVirtualModule } from "./loadVirtualModule.js";
async function getStartManifest(opts) {
  const { tsrStartManifest } = await loadVirtualModule(
    VIRTUAL_MODULES.startManifest
  );
  const startManifest = tsrStartManifest();
  const rootRoute = startManifest.routes[rootRouteId] = startManifest.routes[rootRouteId] || {};
  rootRoute.assets = rootRoute.assets || [];
  if (process.env.NODE_ENV === "development" && !process.env.TSS_CLIENT_ENTRY) {
    throw new Error(
      "tanstack/start-server-core: TSS_CLIENT_ENTRY must be defined in your environment for getStartManifest()"
    );
  }
  if (process.env.NODE_ENV === "development") {
    const clientEntry = joinPaths([opts.basePath, process.env.TSS_CLIENT_ENTRY]);
    const script = `${globalThis.TSS_INJECTED_HEAD_SCRIPTS ? globalThis.TSS_INJECTED_HEAD_SCRIPTS + "; " : ""}import('${clientEntry}')`;
    rootRoute.assets.push({
      tag: "script",
      attrs: {
        type: "module",
        suppressHydrationWarning: true,
        async: true
      },
      children: script
    });
  }
  const manifest = {
    ...startManifest,
    routes: Object.fromEntries(
      Object.entries(startManifest.routes).map(([k, v]) => {
        const { preloads, assets } = v;
        return [
          k,
          {
            preloads,
            assets
          }
        ];
      })
    )
  };
  return manifest;
}
export {
  getStartManifest
};
